airflow-core/src/airflow/models/dagrun.py

```diff
@@ -1217,6 +1218,10 @@ def recalculate(self) -> _UnfinishedStates:
                     msg="success",
                 )
 
+            if (deadline := dag.deadline) and isinstance(deadline.reference, DeadlineReference.TYPES.DAGRUN):
+                # The dagrun has succeeded, so the deadline is no longer needed.
+                Deadline.remove_deadlines(session=session, conditions={DagRun.run_id: self.run_id})
```

Comment:
Why are `conditions` being exposed here? Seems just a tiny bit strange. Shouldn't this just take a param `run_id=self.run_id` and the conditions can be built within the function? Not sure why that's leaking out to the caller.

Reply:
Yeah that's fair. You could have different methods for each use case which assemble the correct conditions internally and then call some shared code for the actually processing of the conditions (basically what you have now). But, very fair if you want to push back on that. Not much internal logic is leaking here so it's somewhat of an academic nit.

Reply:
Yeah, I had considered using kwargs and filtering them out but it loked and felt better explicitly using a parameter, and I feel the name itself fits, too. The method call reads pretty straight forward as "prune deadlines using this session which meet these conditions".  If you're cool with it, I'll resolve this one.